test/core/test_task_launcher.py

```diff
@@ -88,6 +103,37 @@ def test_create_launch_expire_assignments(self):
         for assignment in launcher.assignments:
             self.assertEqual(assignment.get_status(), AssignmentState.EXPIRED)
 
+    def test_launch_assignments_with_concurrent_unit_cap(self):
+        """Initialize a launcher on a task run, then create the assignments"""
+        cap_values = [1, 2, 3, 4, 5]
+        for max_num_units in cap_values:
+            mock_data_array = self.get_mock_assignment_data_array()
+            launcher = TaskLauncher(
+                self.db,
+                self.task_run,
+                mock_data_array,
+                max_num_concurrent_units=max_num_units,
+            )
+            launcher.launched_units = LimitedDict(launcher.max_num_concurrent_units)
+            launcher.create_assignments()
+            launcher.launch_units("dummy-url:3000")
+
+            start_time = time.time()
+            while set([u.get_status() for u in launcher.units]) != {
+                AssignmentState.COMPLETED
+            }:
+                for unit in launcher.units:
+                    if unit.get_status() == AssignmentState.LAUNCHED:
+                        unit.set_db_status(AssignmentState.COMPLETED)
+                    time.sleep(0.1)
+                self.assertEqual(launcher.launched_units.exceed_limit, False)
+                curr_time = time.time()
+                if curr_time - start_time > MAX_WAIT_TIME_UNIT_LAUNCH:
```

Comment:
this should be an assert, rather than a break. Otherwise the test will pass even if this is exceeded.